R/aes-evaluation.r

```diff
@@ -2,39 +2,105 @@
+#' # used as the `mapping` argument in layers.
+#'
+#' @param x <[`data-masking`][rlang::topic-data-mask]> An aesthetic expression
+#'   using variables calculated by the stat (`after_stat()`) or layer aesthetics
+#'   (`after_scale()`).
+#' @param start <[`data-masking`][rlang::topic-data-mask]> An aesthetic
+#'   expression using variables from the layer data.
+#' @param after_stat <[`data-masking`][rlang::topic-data-mask]> An aesthetic
+#'   expression using variables calculated by the stat.
+#' @param after_scale <[`data-masking`][rlang::topic-data-mask]> An aesthetic
+#'   expression using layer aesthetics.
+#'
+#' @note
 #' `after_stat()` replaces the old approaches of using either `stat()` or
 #' surrounding the variable names with `..`.
 #'
-#' @note Evaluation after stat transformation will have access to the
-#' variables calculated by the stat, not the original mapped values. Evaluation
-#' after scaling will only have access to the final aesthetics of the layer
-#' (including non-mapped, default aesthetics). The original layer data can only
-#' be accessed at the first stage.
+#' @section Staging:
+#' Below follows an overview of the three stages of evaluation and how aesthetic
+#' evaluation can be controlled.
+#'
+#' ## Stage 1: direct input
+#'   The default is to map at the beginning, using the layer data provided by
+#'   the user. If you want to map directly from the layer data you should not do
+#'   anything special. This is the only stage where the original layer data can
+#'   be accessed.
+#'
+#' ```{r direct_input, eval = FALSE}
+#' # 'x' and 'y' are mapped directly
+#' ggplot(mtcars) + geom_point(aes(x = mpg, y = disp))
+#' ```
+#'
+#' ## Stage 2: after stat transformation
+#'   The second stage is after the data has been transformed by the layer
+#'   stat. The most common example of mapping from stat transformed data is the
+#'   height of bars in [geom_histogram()]: the height does not come from a
+#'   variable in the underlying data, but is instead mapped to the `count`
+#'   computed by [stat_bin()]. In order to map from stat transformed data you
+#'   should use the `after_stat()` function to flag that evaluation of the
+#'   aesthetic mapping should be postponed until after stat transformation.
+#'   Evaluation after stat transformation will have access to the variables
+#'   calculated by the stat, not the original mapped values. The 'computed
+#'   variables' section in each stat lists which variables are available to
+#'   access.
+#'
+#' ```{r after_stat_transformation, eval = FALSE}
+#' # The 'y' values for the histogram are computed by the stat
+#' ggplot(faithful, aes(x = waiting)) +
+#'   geom_histogram()
+#'
+#' # Choosing a different computed variable to display, matching up the
+#' # histogram with the density plot
+#' ggplot(faithful, aes(x = waiting)) +
+#'   geom_histogram(aes(y = after_stat(density))) +
+#'   geom_density()
+#' ```
+#'
```

Comment:
Might be worth mentioning deprecated `..density..` style.

Reply:
I've more explicitly put this in the `@note` now.